tests/erc721/erc721_test.go

```diff
@@ -162,6 +171,289 @@ func TestOpenSeaProxyApproval(t *testing.T) {
+			wantPreApproved: false,
+			wantEvents:      nil,
+		},
+		{
+			name:            "mint with proxy",
+			from:            -1,
+			to:              tokenReceiver,
+			hasProxy:        true,
+			wantPreApproved: true,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+			},
+		},
+		{
+			name:            "transfer without proxy",
+			from:            tokenOwner,
+			to:              tokenReceiver,
+			hasProxy:        false,
+			wantPreApproved: false,
+			wantEvents:      nil,
+		},
+		{
+			name:            "transfer with proxy",
+			from:            tokenOwner,
+			to:              tokenReceiver,
+			hasProxy:        true,
+			wantPreApproved: true,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+			},
+		},
+		{
+			name:            "burn without proxy",
+			from:            tokenOwner,
+			to:              -1,
+			hasProxy:        false,
+			wantPreApproved: false,
+			wantEvents:      nil,
+		},
+		{
+			name:            "burn with proxy",
+			from:            tokenOwner,
+			to:              -1,
+			hasProxy:        false,
+			wantPreApproved: false,
+			wantEvents:      nil,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			sim, nft, filter := deploy(t)
+			start := sim.BlockNumber().Uint64()
+
+			toAddr := common.HexToAddress("0x0")
+			if tt.to >= 0 {
+				toAddr = sim.Addr(tt.to)
+			}
+
+			if tt.hasProxy {
+				openseatest.SetProxyTB(t, sim, toAddr, sim.Addr(proxy))
+			}
+
+			if tt.from == -1 {
+				sim.Must(t, "Mint(%d)", notExists)(nft.Mint(sim.Acc(tt.to), big.NewInt(notExists)))
+			} else if tt.to == -1 {
+				sim.Must(t, "Burn(%d)", exists)(nft.Burn(sim.Acc(tt.from), big.NewInt(exists)))
+			} else {
+				sim.Must(t, "TransferFrom(%d)", exists)(nft.TransferFrom(sim.Acc(tt.from), sim.Addr(tt.from), toAddr, big.NewInt(exists)))
+			}
+
+			testApprovalForAllEvents(t, filter, start, tt.wantEvents)
+
+			{
+				got, err := nft.IsApprovedForAll(nil, toAddr, sim.Addr(proxy))
+				if err != nil || got != tt.wantPreApproved {
+					t.Errorf("IsApprovedForAll([owner], [proxy]) got %t, err = %v; want %t, nil err", got, err, tt.wantPreApproved)
+				}
+			}
+		})
+	}
+}
+
+func TestOpenSeaProxyPreApprovalOptInOut(t *testing.T) {
+	sim, _, _ := deploy(t)
+
+	tests := []struct {
+		name         string
+		mints        bool
+		hasProxy     bool
+		wantApproved bool
+		wantEvents   []*ERC721ApprovalForAll
+	}{
+		{
+			name:         "opt in with proxy after mint",
+			mints:        true,
+			hasProxy:     true,
+			wantApproved: true,
+			wantEvents: []*ERC721ApprovalForAll{
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
+				{
+					Owner:    sim.Addr(tokenReceiver),
+					Operator: sim.Addr(proxy),
+					Approved: true,
+				},
```

Comment:
Shouldn't this be a no-op when they call `setApprovalForAll()` as their state is already active by default?

Reply:
Here I tried to mimic the standard behavior of OZ's ERC721, which also emits multiple events if `setApprovalForAll` is called repeatedly.